File: artifacts/lacuna-mobile/lib/dataset.ts
import rawDataset from "@/data/dataset.json";

export interface Company {
  id: string;
  name: string;
  sector: string;
  stage: string;
  founded: number;
  hq: string;
  description: string;
  lastKnownValuation?: number;
  valuationSource?: string;
  totalFunding?: number;
  sources?: string[];
}

export interface Acquirer {
  id: string;
  name: string;
  ticker?: string;
  sector: string;
  hq: string;
}

export interface Acquisition {
  id: string;
  targetId: string;
  acquirerId: string;
  targetName: string;
  acquirerName: string;
  announcedDate: string;
  closedDate?: string;
  dealValue?: number;
  dealValueNote?: string;
  dealType: string;
  source: string;
  strategicRationale: string;
}

export interface Dataset {
  provenance: {
    lastUpdated: string;
    datasetVersion?: string;
    sources: string[];
    notes: string[];
    purpose: string;
    disclaimer: string;
  };
  companies: Company[];
  acquirers: Acquirer[];
  acquisitions: Acquisition[];
}

// @ts-ignore
const dataset = rawDataset as Dataset;

export const companies: Company[] = dataset.companies;
export const acquirers: Acquirer[] = dataset.acquirers;
export const acquisitions: Acquisition[] = dataset.acquisitions;
export const provenance = dataset.provenance;

export function getTotalDealValue(): number {
  return acquisitions.reduce((sum, a) => sum + (a.dealValue ?? 0), 0);
}

export function getDealsByYear(): Array<{ year: number; count: number }> {
  const map = new Map<number, number>();
  for (const acq of acquisitions) {
    const year = parseInt(acq.announcedDate.slice(0, 4), 10);
    if (!isNaN(year)) map.set(year, (map.get(year) ?? 0) + 1);
  }
  return Array.from(map.entries())
    .map(([year, count]) => ({ year, count }))
    .sort((a, b) => a.year - b.year);
}

export function getSectorCounts(): Array<{ sector: string; count: number }> {
  const map = new Map<string, number>();
  for (const c of companies) {
    map.set(c.sector, (map.get(c.sector) ?? 0) + 1);
  }
  return Array.from(map.entries())
    .map(([sector, count]) => ({ sector, count }))
    .sort((a, b) => b.count - a.count);
}

export function getAcquirerDealCounts(): Array<{ name: string; count: number; id: string }> {
  const map = new Map<string, { name: string; count: number }>();
  for (const acq of acquisitions) {
    const existing = map.get(acq.acquirerId);
    if (existing) existing.count++;
    else map.set(acq.acquirerId, { name: acq.acquirerName, count: 1 });
  }
  return Array.from(map.entries())
    .map(([id, { name, count }]) => ({ id, name, count }))
    .sort((a, b) => b.count - a.count);
}

export function formatDealValue(value?: number): string {
  if (!value) return "Undisclosed";
  if (value >= 1000) return `$${(value / 1000).toFixed(1)}B`;
  return `$${value}M`;
}

File: services/python-api/app/dataset.py
import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

DatasetResource = Literal["companies", "acquisitions", "acquirers", "all"]

GENOMICS_KEYWORD = re.compile(
    r"genomic|genome|sequenc|brca|biomarker|hereditary|carrier screening|cgp|"
    r"profiling|variant|exome|oncotype|pcos|sickle|hbb|lynch|lupus|hla|"
    r"palb2|chek2|dennd1a|fshr",
    re.IGNORECASE,
)


class Provenance(BaseModel):
    lastUpdated: str
    datasetVersion: str | None = None
    sources: list[str]
    notes: list[str] = Field(default_factory=list)
    purpose: str
    disclaimer: str


class Company(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    name: str
    sector: str = "Unknown"
    stage: str = ""
    founded: int = 0
    hq: str = ""
    description: str = ""
    lastKnownValuation: float | None = None
    valuationSource: str | None = None
    totalFunding: float | None = None
    sources: list[str] | None = None
    evidenceClass: str | None = None


class Acquirer(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    name: str
    ticker: str | None = None
    sector: str | None = None
    hq: str = ""
    type: str | None = None
    description: str | None = None


class Acquisition(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    targetId: str
    acquirerId: str
    targetName: str
    acquirerName: str
    announcedDate: str
    dealType: str
    source: str
    strategicRationale: str = ""
    closedDate: str | None = None
    dealValue: float | None = None
    dealValueNote: str | None = None


class VerifiedDataset(BaseModel):
    provenance: Provenance
    companies: list[Company]
    acquirers: list[Acquirer]
    acquisitions: list[Acquisition]


class DatasetMeta(BaseModel):
    resource: DatasetResource
    limit: int
    offset: int
    sector: str | None = None
    genomics: bool = False
    total: dict[str, int]


class DatasetSlice(BaseModel):
    provenance: Provenance
    companies: list[Company]
    acquirers: list[Acquirer]
    acquisitions: list[Acquisition]
    meta: DatasetMeta


def is_genomics_relevant_company(company: Company) -> bool:
    if company.sector == "Diagnostics":
        return True
    haystack = f"{company.name} {company.description}"
    return bool(GENOMICS_KEYWORD.search(haystack))


def _paginate(items: list[Any], limit: int, offset: int) -> list[Any]:
    return items[offset : offset + limit]


@lru_cache
def load_verified_dataset(dataset_path: str) -> VerifiedDataset:
    path = Path(dataset_path)
    with path.open(encoding="utf-8") as handle:
        payload = json.load(handle)
    return VerifiedDataset.model_validate(payload)


def slice_verified_dataset(
    dataset: VerifiedDataset,
    *,
    resource: DatasetResource = "all",
    limit: int = 50,
    offset: int = 0,
    sector: str | None = None,
    genomics: bool = False,
) -> DatasetSlice:
    companies_by_id = {company.id: company for company in dataset.companies}
    companies = list(dataset.companies)
    acquisitions = list(dataset.acquisitions)

    if sector:
        companies = [c for c in companies if c.sector == sector]
        sector_ids = {c.id for c in companies}
        acquisitions = [d for d in acquisitions if d.targetId in sector_ids]

    if genomics:
        companies = [c for c in companies if is_genomics_relevant_company(c)]
        acquisitions = [
            d
            for d in acquisitions
            if (target := companies_by_id.get(d.targetId)) is not None
            and is_genomics_relevant_company(target)
        ]

    totals = {
        "companies": len(companies),
        "acquisitions": len(acquisitions),
        "acquirers": len(dataset.acquirers),
    }

    sliced_companies = (
        _paginate(companies, limit, offset)
        if resource in ("all", "companies")
        else []
    )
    sliced_acquisitions = (
        _paginate(acquisitions, limit, offset)
        if resource in ("all", "acquisitions")
        else []
    )
    sliced_acquirers = (
        _paginate(dataset.acquirers, limit, offset)
        if resource in ("all", "acquirers")
        else []
    )

    return DatasetSlice(
        provenance=dataset.provenance,
        companies=sliced_companies,
        acquirers=sliced_acquirers,
        acquisitions=sliced_acquisitions,
        meta=DatasetMeta(
            resource=resource,
            limit=limit,
            offset=offset,
            sector=sector,
            genomics=genomics,
            total=totals,
        ),
    )
